### PR TITLE
Fix NullEquals binary op regression

### PR DESCRIPTION
**Pull request by karthikeyann**
partial fix to https://github.com/rapidsai/cudf/issues/9795
move `NullEquals` binary op to separate file instead of fused single kernel in file `equality_ops.cu`
Possible cause for slow down is register spilling. This PR separated `NullEquals` to separate kernel.

### CODE REVIEW

**Comment by karthikeyann** [comment]
Benchmark results compared (OLD vs this-PR) for NULL_EQUALS binary op on all types.
```
Comparing ./NULL_EQUALS_22.02.json to ./FIX_NULL_EQUALS_22.02.json
Benchmark                                                                                                                       Time             CPU      Time Old      Time New       CPU Old       CPU New
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
COMPILED_BINARYOP<int8_t, int8_t, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                                -0.6290         -0.3524            21             8            42            27
COMPILED_BINARYOP<int8_t, int8_t, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                               -0.8153         -0.6147            51             9            72            28
COMPILED_BINARYOP<int8_t, int8_t, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                              -0.9509         -0.9165           440            22           463            39
COMPILED_BINARYOP<int8_t, int8_t, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                             -0.9683         -0.9643          4218           134          4266           152
COMPILED_BINARYOP<int8_t, int8_t, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                            -0.9697         -0.9709         42075          1275         44595          1296
COMPILED_BINARYOP<int16_t, int16_t, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                              -0.6383         -0.3636            21             8            42            27
COMPILED_BINARYOP<int16_t, int16_t, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                             -0.8147         -0.6276            53            10            75            28
COMPILED_BINARYOP<int16_t, int16_t, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                            -0.9446         -0.9130           443            25           468            41
COMPILED_BINARYOP<int16_t, int16_t, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                           -0.9649         -0.9615          4293           151          4365           168
COMPILED_BINARYOP<int16_t, int16_t, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                          -0.9659         -0.9672         42634          1455         44965          1476
COMPILED_BINARYOP<int32_t, int32_t, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                              -0.6079         -0.3494            21             8            42            27
COMPILED_BINARYOP<int32_t, int32_t, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                             -0.8029         -0.6204            54            11            76            29
COMPILED_BINARYOP<int32_t, int32_t, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                            -0.9330         -0.9032           452            30           480            46
COMPILED_BINARYOP<int32_t, int32_t, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                           -0.9569         -0.9533          4331           187          4361           204
COMPILED_BINARYOP<int32_t, int32_t, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                          -0.9587         -0.9605         43200          1786         45780          1808
COMPILED_BINARYOP<int64_t, int64_t, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                              -0.6017         -0.3497            22             9            43            28
COMPILED_BINARYOP<int64_t, int64_t, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                             -0.7726         -0.5931            54            12            75            31
COMPILED_BINARYOP<int64_t, int64_t, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                            -0.9063         -0.8775           467            44           491            60
COMPILED_BINARYOP<int64_t, int64_t, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                           -0.9337         -0.9309          4531           300          4593           318
COMPILED_BINARYOP<int64_t, int64_t, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                          -0.9356         -0.9386         44903          2891         47552          2920
COMPILED_BINARYOP<uint8_t, uint8_t, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                              -0.6388         -0.3695            22             8            43            27
COMPILED_BINARYOP<uint8_t, uint8_t, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                             -0.8055         -0.6034            50            10            71            28
COMPILED_BINARYOP<uint8_t, uint8_t, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                            -0.9481         -0.9132           436            23           458            40
COMPILED_BINARYOP<uint8_t, uint8_t, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                           -0.9660         -0.9622          4248           145          4311           163
COMPILED_BINARYOP<uint8_t, uint8_t, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                          -0.9671         -0.9685         42126          1384         44614          1405
COMPILED_BINARYOP<uint16_t, uint16_t, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                            -0.6361         -0.3651            22             8            43            27
COMPILED_BINARYOP<uint16_t, uint16_t, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                           -0.8105         -0.6174            52            10            74            28
COMPILED_BINARYOP<uint16_t, uint16_t, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                          -0.9432         -0.9122           445            25           472            41
COMPILED_BINARYOP<uint16_t, uint16_t, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                         -0.9632         -0.9600          4298           158          4377           175
COMPILED_BINARYOP<uint16_t, uint16_t, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                        -0.9643         -0.9659         42657          1521         45210          1543
COMPILED_BINARYOP<uint32_t, uint32_t, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                            -0.6189         -0.3635            22             8            43            28
COMPILED_BINARYOP<uint32_t, uint32_t, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                           -0.7946         -0.6074            53            11            74            29
COMPILED_BINARYOP<uint32_t, uint32_t, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                          -0.9327         -0.9030           452            30           480            47
COMPILED_BINARYOP<uint32_t, uint32_t, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                         -0.9569         -0.9534          4348           187          4389           205
COMPILED_BINARYOP<uint32_t, uint32_t, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                        -0.9584         -0.9603         43230          1798         45893          1820
COMPILED_BINARYOP<uint64_t, uint64_t, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                            -0.6074         -0.3612            22             9            44            28
COMPILED_BINARYOP<uint64_t, uint64_t, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                           -0.7784         -0.6038            55            12            77            31
COMPILED_BINARYOP<uint64_t, uint64_t, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                          -0.9066         -0.8783           468            44           493            60
COMPILED_BINARYOP<uint64_t, uint64_t, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                         -0.9340         -0.9315          4536           299          4615           316
COMPILED_BINARYOP<uint64_t, uint64_t, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                        -0.9356         -0.9387         44920          2892         47635          2921
COMPILED_BINARYOP<float, float, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                                  -0.6209         -0.3685            22             8            44            28
COMPILED_BINARYOP<float, float, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                                 -0.7991         -0.6122            53            11            75            29
COMPILED_BINARYOP<float, float, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                                -0.9330         -0.9033           455            30           483            47
COMPILED_BINARYOP<float, float, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                               -0.9568         -0.9536          4351           188          4413           205
COMPILED_BINARYOP<float, float, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                              -0.9583         -0.9603         43219          1801         45919          1823
COMPILED_BINARYOP<double, double, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                                -0.6002         -0.3562            22             9            44            28
COMPILED_BINARYOP<double, double, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                               -0.7737         -0.5990            54            12            77            31
COMPILED_BINARYOP<double, double, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                              -0.9065         -0.8786           469            44           496            60
COMPILED_BINARYOP<double, double, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                             -0.9339         -0.9312          4530           299          4595           316
COMPILED_BINARYOP<double, double, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                            -0.9356         -0.9388         44921          2892         47759          2921
COMPILED_BINARYOP<bool, bool, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                                    -0.6115         -0.3561            22             9            43            28
COMPILED_BINARYOP<bool, bool, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                                   -0.7902         -0.6043            51            11            74            29
COMPILED_BINARYOP<bool, bool, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                                  -0.9348         -0.9016           442            29           466            46
COMPILED_BINARYOP<bool, bool, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                                 -0.9536         -0.9500          4272           198          4339           217
COMPILED_BINARYOP<bool, bool, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                                -0.9550         -0.9572         42469          1909         45127          1933
COMPILED_BINARYOP<timestamp_D, timestamp_D, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                      -0.6145         -0.3551            21             8            42            27
COMPILED_BINARYOP<timestamp_D, timestamp_D, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                     -0.7905         -0.6009            52            11            74            29
COMPILED_BINARYOP<timestamp_D, timestamp_D, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                    -0.9315         -0.8999           439            30           462            46
COMPILED_BINARYOP<timestamp_D, timestamp_D, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                   -0.9566         -0.9532          4281           186          4329           203
COMPILED_BINARYOP<timestamp_D, timestamp_D, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                  -0.9585         -0.9599         42722          1774         44762          1797
COMPILED_BINARYOP<timestamp_s, timestamp_s, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                      -0.5978         -0.3481            21             9            43            28
COMPILED_BINARYOP<timestamp_s, timestamp_s, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                     -0.7796         -0.6032            54            12            76            30
COMPILED_BINARYOP<timestamp_s, timestamp_s, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                    -0.9063         -0.8773           464            43           487            60
COMPILED_BINARYOP<timestamp_s, timestamp_s, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                   -0.9339         -0.9310          4519           299          4570           316
COMPILED_BINARYOP<timestamp_s, timestamp_s, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                  -0.9356         -0.9387         44813          2887         47588          2916
COMPILED_BINARYOP<timestamp_ms, timestamp_ms, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                    -0.6125         -0.3484            22             9            42            28
COMPILED_BINARYOP<timestamp_ms, timestamp_ms, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                   -0.7791         -0.6016            54            12            76            30
COMPILED_BINARYOP<timestamp_ms, timestamp_ms, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                  -0.9064         -0.8778           466            44           492            60
COMPILED_BINARYOP<timestamp_ms, timestamp_ms, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                 -0.9336         -0.9312          4517           300          4583           316
COMPILED_BINARYOP<timestamp_ms, timestamp_ms, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                -0.9355         -0.9388         44795          2888         47615          2916
COMPILED_BINARYOP<timestamp_us, timestamp_us, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                    -0.6154         -0.3677            22             9            44            28
COMPILED_BINARYOP<timestamp_us, timestamp_us, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                   -0.7816         -0.6092            55            12            78            30
COMPILED_BINARYOP<timestamp_us, timestamp_us, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                  -0.9071         -0.8794           469            44           497            60
COMPILED_BINARYOP<timestamp_us, timestamp_us, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                 -0.9338         -0.9311          4518           299          4579           316
COMPILED_BINARYOP<timestamp_us, timestamp_us, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                -0.9356         -0.9389         44891          2889         47702          2916
COMPILED_BINARYOP<timestamp_ns, timestamp_ns, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                    -0.6124         -0.3653            22             9            44            28
COMPILED_BINARYOP<timestamp_ns, timestamp_ns, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                   -0.7802         -0.6027            55            12            77            30
COMPILED_BINARYOP<timestamp_ns, timestamp_ns, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                  -0.9066         -0.8782           467            44           493            60
COMPILED_BINARYOP<timestamp_ns, timestamp_ns, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                 -0.9338         -0.9312          4524           299          4592           316
COMPILED_BINARYOP<timestamp_ns, timestamp_ns, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                -0.9356         -0.9388         44903          2891         47682          2920
COMPILED_BINARYOP<duration_D, duration_D, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                        -0.6218         -0.3619            22             8            43            27
COMPILED_BINARYOP<duration_D, duration_D, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                       -0.7979         -0.6122            52            11            75            29
COMPILED_BINARYOP<duration_D, duration_D, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                      -0.9319         -0.9008           442            30           466            46
COMPILED_BINARYOP<duration_D, duration_D, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                     -0.9571         -0.9538          4299           185          4365           202
COMPILED_BINARYOP<duration_D, duration_D, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                    -0.9586         -0.9606         42729          1770         45433          1792
COMPILED_BINARYOP<duration_s, duration_s, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                        -0.6017         -0.3587            22             9            44            28
COMPILED_BINARYOP<duration_s, duration_s, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                       -0.7795         -0.6048            54            12            77            30
COMPILED_BINARYOP<duration_s, duration_s, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                      -0.9069         -0.8786           467            43           492            60
COMPILED_BINARYOP<duration_s, duration_s, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                     -0.9338         -0.9310          4517           299          4576           316
COMPILED_BINARYOP<duration_s, duration_s, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                    -0.9356         -0.9388         44861          2887         47657          2915
COMPILED_BINARYOP<duration_ms, duration_ms, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                      -0.6002         -0.3504            21             9            43            28
COMPILED_BINARYOP<duration_ms, duration_ms, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                     -0.7765         -0.5970            54            12            75            30
COMPILED_BINARYOP<duration_ms, duration_ms, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                    -0.9068         -0.8784           466            43           492            60
COMPILED_BINARYOP<duration_ms, duration_ms, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                   -0.9339         -0.9314          4521           299          4597           316
COMPILED_BINARYOP<duration_ms, duration_ms, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                  -0.9356         -0.9388         44828          2888         47668          2916
COMPILED_BINARYOP<duration_us, duration_us, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                      -0.6062         -0.3560            22             9            43            28
COMPILED_BINARYOP<duration_us, duration_us, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                     -0.7787         -0.6023            54            12            76            30
COMPILED_BINARYOP<duration_us, duration_us, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                    -0.9065         -0.8780           466            44           491            60
COMPILED_BINARYOP<duration_us, duration_us, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                   -0.9338         -0.9311          4528           300          4597           317
COMPILED_BINARYOP<duration_us, duration_us, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                  -0.9356         -0.9388         44888          2890         47700          2918
COMPILED_BINARYOP<duration_ns, duration_ns, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                      -0.5985         -0.3534            22             9            43            28
COMPILED_BINARYOP<duration_ns, duration_ns, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                     -0.7780         -0.6015            55            12            77            31
COMPILED_BINARYOP<duration_ns, duration_ns, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                    -0.9056         -0.8763           464            44           486            60
COMPILED_BINARYOP<duration_ns, duration_ns, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                   -0.9336         -0.9307          4513           300          4563           316
COMPILED_BINARYOP<duration_ns, duration_ns, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                  -0.9356         -0.9387         44932          2892         47665          2921
COMPILED_BINARYOP<decimal32, decimal32, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                          -0.6150         -0.3618            23             9            44            28
COMPILED_BINARYOP<decimal32, decimal32, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                         -0.7978         -0.6165            54            11            76            29
COMPILED_BINARYOP<decimal32, decimal32, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                        -0.9321         -0.9018           458            31           481            47
COMPILED_BINARYOP<decimal32, decimal32, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                       -0.9566         -0.9533          4450           193          4500           210
COMPILED_BINARYOP<decimal32, decimal32, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                      -0.9584         -0.9604         44398          1845         47157          1867
COMPILED_BINARYOP<decimal64, decimal64, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                          -0.6012         -0.3571            22             9            44            28
COMPILED_BINARYOP<decimal64, decimal64, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                         -0.7779         -0.6010            55            12            77            31
COMPILED_BINARYOP<decimal64, decimal64, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                        -0.9079         -0.8801           478            44           503            60
COMPILED_BINARYOP<decimal64, decimal64, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                       -0.9354         -0.9327          4636           300          4704           316
COMPILED_BINARYOP<decimal64, decimal64, bool, cudf::binary_operator::NULL_EQUALS>/100000000/manual_time                      -0.9370         -0.9398         45967          2894         48503          2922
COMPILED_BINARYOP<decimal128, decimal128, bool, cudf::binary_operator::NULL_EQUALS>/10000/manual_time                        -0.5938         -0.3668            23             9            45            29
COMPILED_BINARYOP<decimal128, decimal128, bool, cudf::binary_operator::NULL_EQUALS>/100000/manual_time                       -0.7440         -0.5837            59            15            79            33
COMPILED_BINARYOP<decimal128, decimal128, bool, cudf::binary_operator::NULL_EQUALS>/1000000/manual_time                      -0.8619         -0.8400           514            71           543            87
COMPILED_BINARYOP<decimal128, decimal128, bool, cudf::binary_operator::NULL_EQUALS>/10000000/manual_time                     -0.8888         -0.8875          4991           555          5078           571

```

**Comment by jrhemstad** [comment]
I have a hard time believing that register pressure caused a 30x slowdown in such a trivially compute bound kernel. Something else must be going on here.

**Comment by robertmaynard** [comment]
> I have a hard time believing that register pressure caused a 30x slowdown in such a trivially compute bound kernel. Something else must be going on here.

Doesn't nvcc have a time limit for optimizing? Moving to multiple TU's would give us more time per kernel for optimization.
If int128 increased the number of kernels in the old TU, we could be getting less optimization time per kernel

File: cpp/src/binaryop/compiled/equality_ops.cu
Comment (bdice) [inline cpp/src/binaryop/compiled/equality_ops.cu:39]: Does it still make sense to dispatch `EQUAL` and `NOT_EQUAL` in this way? The code claims just above this line that
```
// Similar enabled template types should go together (better performance)
```
but that seems not to be true for the case of `NULL_EQUALS` (which is why this PR was created).

Why would this affect `NULL_EQUALS` but not the other dispatches? Additional handling of bitmask data that doesn't occur for the other two?

Comment (karthikeyann) [reply cpp/src/binaryop/compiled/equality_ops.cu:39]: It still might affect these 2 dispatch as well. will benchmark these too.
When I worked on compiled binary ops, I tried several combinations of operators to fuse. 
Combining Less, LessEqual,  Greater, GreaterEqual faced exact same performance issue shown here. Hence, this quick fix made sense.

Fusing was done for couple of reasons - 1. compile time, 2. performance. Of all the combinations I tried, this only had better  performance and lesser compile time, though performance benefit was much less.

**Review by revans2 — APPROVED** [review]
I was able to confirm that this fixed the issue we were seeing that triggered the original BUG. I still need to look at what other performance differences there still are, but I think this is good enough to unblock us.

**Comment by karthikeyann** [comment]
cuobjdump resource usage

equality_ops.cu (EQUALS, NOT_EQUAL, NULL_EQUALS):
   REG:42 STACK:208 SHARED:0 LOCAL:0 CONSTANT[0]:592 CONSTANT[2]:4832 TEXTURE:0 SURFACE:0 SAMPLER:0
equality_ops.cu ((EQUALS, NOT_EQUAL):
   REG:40 STACK:0 SHARED:0 LOCAL:0 CONSTANT[0]:592 CONSTANT[2]:3208 TEXTURE:0 SURFACE:0 SAMPLER:0
NullEquals.cu (NULL_EQUALS):
  REG:42 STACK:0 SHARED:0 LOCAL:0 CONSTANT[0]:576 CONSTANT[2]:1600 TEXTURE:0 SURFACE:0 SAMPLER:0

I could see difference between constant memory only.

File: cpp/src/binaryop/compiled/equality_ops.cu
Comment (bdice) [reply cpp/src/binaryop/compiled/equality_ops.cu:39]: > cuobjdump resource usage
> 
> equality_ops.cu (EQUALS, NOT_EQUAL, NULL_EQUALS): REG:42 STACK:208 SHARED:0 LOCAL:0 CONSTANT[0]:592 CONSTANT[2]:4832 TEXTURE:0 SURFACE:0 SAMPLER:0 equality_ops.cu ((EQUALS, NOT_EQUAL): REG:40 STACK:0 SHARED:0 LOCAL:0 CONSTANT[0]:592 CONSTANT[2]:3208 TEXTURE:0 SURFACE:0 SAMPLER:0 NullEquals.cu (NULL_EQUALS): REG:42 STACK:0 SHARED:0 LOCAL:0 CONSTANT[0]:576 CONSTANT[2]:1600 TEXTURE:0 SURFACE:0 SAMPLER:0
> 
> I could see difference between constant memory and STACK. Looking into ptx and profiling too.

My suspicion from those values is that splitting the `EQUAL` / `NOT_EQUAL` dispatcher won't improve performance any further, and that it is probably fine to leave as-is (go ahead and post the results if you choose to test performance with this split apart).

From what I saw with the AST code, I would guess that the stack frame with `NULL_EQUALS` is killing the performance due to local memory transfer.

**Review by bdice — APPROVED** [review]
Approved: the changeset looks good, we need to push this fix through for the 21.12 release, and I don't expect further improvements from splitting that dispatcher based on @karthikeyann's `cuobjdump` results.

Great work as always, @karthikeyann!

**Comment by revans2** [comment]
There still are a number of performance regressions that I am seeing in other binops. 

Comparison binops for decimal values, i.e. EQUAL, LESS,  GREATER_EQUAL are still much slower than they were before, at lest 3x slower for an end to end query.

Math operations for int types INT64, INT32, INT16, and INT8 are about 2x slower than they were before, i.e. ADD, SUB, MUL, and DIV. Oddly MOD is not impacted by this.  I could not easily test if this also impacted decimal operations too, just because of how Spark works.

The comparisons I am not concerned about because they appear to be isolated to just Decimal, but the math operations on INTs are a little concerning.  @sameerz @jlowe @tgravescs do you think that these would be blockers for 21.12? I personally don't think that they are but it is hard to tell because the current problem was so big it could be masking other thing.

**Review by davidwendt — APPROVED** [review]

**Comment by jrhemstad** [comment]
https://github.com/rapidsai/cudf/issues/9813 this is certainly part of the problem here.

**Comment by jlowe** [comment]
The decimal slowdowns are probably OK for 21.12 from the RAPIDS Accelerator perspective since it doesn't support decimals by default in 21.10 so there's no previous default GPU baseline for most users to notice.  The integer arithmetic performance loss is unfortunate but probably tolerable for 21.12 if it's "only" 2X slower.  We haven't seen integer arithmetic ops appear on our profile radar before, and I suspect even at 2X it will be a minimal performance impact for most queries.  However we should do a full NDS benchmark run comparison to 21.10 with this fix to verify we're close enough to 21.10 performance on end-to-end query times to unblock the 21.12 release.

**Comment by robertmaynard** [comment]
So if we look at the regression in nsight compute compared to the fix as proposed in this PR, I see
a couple of things indicating that our jumbo kernel for binary_ops is not optimizing well.

The primary indicator is the introduction of `LDL` and `CALL` instructions compared to the baseline, and secondary is the significantly higher amount of `IMAD` instructions. 

Regression:
SOL Regression
Instruction Count Regression


This PR:
SOL Correction
Instruction Count Regression

**Comment by jrhemstad** [comment]
> So if we look at the regression in nsight compute compared to the fix as proposed in this PR, I see a couple of things indicating that our jumbo kernel for binary_ops is not optimizing well.

I'd look at the memory throughput. My guess is the increased stack frame is causing significantly increased global memory traffic.

**Review by codereport — APPROVED** [review]

**Comment by robertmaynard** [comment]
> > So if we look at the regression in nsight compute compared to the fix as proposed in this PR, I see a couple of things indicating that our jumbo kernel for binary_ops is not optimizing well.
> 
> I'd look at the memory throughput. My guess is the increased stack frame is causing significantly increased global memory traffic.

If we look only at global throughput we see the regression 404 GB vs this PR at 533 but with a vastly higher amount of total memory moved. This PR moves 5m sectors to device memory, while the regression moves 81m.

Regression:
Screenshot from 2021-12-01 13-07-42




This PR:
Screenshot from 2021-12-01 13-05-42

**Comment by jrhemstad** [comment]
> This PR moves 5m sectors to device memory, while the regression moves 81m.

Well there's your problem. So clearly it's a confluence of the optimizer giving up and causing stack frame overhead that dramatically increases global memory traffic.

**Review by robertmaynard — APPROVED** [review]

**Review by jrhemstad — CHANGES_REQUESTED** [review]
I'm going to block merging this until we have some hard numbers on what regressions still remain and how significant they are as well as a more thorough understanding of what caused them.

**Review by jrhemstad — APPROVED** [review]

**Review by ajschmidt8 — APPROVED** [review]

**Comment by raydouglass** [comment]
I believe this was superseded by #9816. Please reopen if that is not the case.

**Pull request closed without merging**
